In [179]:
# Dependencies and Setup
import pandas as pd
import numpy as np

%matplotlib inline
import matplotlib.pyplot as plt

import requests
import json

import time # use for annotating the date of the API pull in graphics using '.now'

import openbrewerydb # a wrapper for accessing the Open Brewery DB API

import gmaps
import gmaps.datasets

import gmaps.geojson_geometries

from matplotlib.cm import viridis
from matplotlib.colors import to_hex

#config
from config import api_key

from opencage.geocoder import OpenCageGeocode
from pprint import pprint


gmaps.configure(api_key=api_key)
# print(api_key) # comment out before saving or uploading to github

In [181]:
brewery_data_df = openbrewerydb.load(state=None)
brewery_data_df.count() # there are approximately 8,000 breweries in the US as of 2019

brewery_type    8029
city            7924
country         8029
id              8029
latitude        4737
longitude       4737
name            8029
phone           8029
postal_code     7924
state           7924
street          8029
tag_list        8029
updated_at      8029
website_url     8029
dtype: int64

#### Note: The count of breweries with latitude and longitude is 4,737 - this will be the (inital) "demo" list of breweries  for mapping
* The brewery_data_df has over 8,000 entries; however, only 4,737 have a lat/lon entry
* A better location choice (one that gets the most brewery locations) is city, state and/or postal_code
* Follow-up work will create missing latitude and longitude from the available city, state, postal_code information for those having an "NaN" which should increase the number of breweries back up to 7,924.
* Is is assumed that the relative density of breweries as would not be depicted on a map would be signficantly different for cities with a larger number of breweries (~ 23 or more).

In [182]:
# View brewery_data_df looking for NaN for latitude and longitude 
brewery_data_df.head(120) # note the brewpub in Prescott Valley, AZ (entry 92) has NaN for latitude and longitude 

brewery_type               city        country    id   latitude  \
0          micro         Birmingham  United States     2  33.524521   
1          micro         Tuscaloosa  United States     4  33.198491   
2          micro         Birmingham  United States    44  33.512849   
3          micro         Huntsville  United States    46  34.727752   
4          micro            Wasilla  United States    55  61.575269   
5          micro          Anchorage  United States    76  61.138489   
6          micro             Tucson  United States    94  32.246737   
7       contract         Scottsdale  United States    98  33.497262   
8        brewpub           Chandler  United States   104  33.305346   
9          micro             Tucson  United States   107  32.201608   
10         micro             Tucson  United States   127  32.250495   
11         micro           Williams  United States   141  35.250028   
12         micro   Lake Havasu City  United States   163  34.468974   
13         micro             Peoria  United States   182  33.668774   
14      regional           Chandler  United States   187  33.303244   
15       brewpub           Surprise  United States   198  33.638221   
16         micro            Phoenix  United States   219  33.516633   
17       brewpub          Paragould  United States   225  36.031636   
18         micro  North Little Rock  United States   235  34.759428   
19         micro        Little Rock  United States   247  34.742845   
20         micro      Mountain Home  United States   253  36.332643   
21      planning             Lowell  United States   264  36.255354   
22       brewpub    San Luis Obispo  United States   281  35.246728   
23         micro          San Diego  United States   286  33.023910   
24         micro          Livermore  United States   305  37.678391   
25       brewpub             Auburn  United States   329  38.896121   
26         micro            Anaheim  United States   333  33.799837   
27       brewpub       Rohnert Park  United States   361  38.364970   
28         micro            Lemoore  United States   376  36.301468   
29         micro      San Francisco  United States   385  37.780655   
..           ...                ...            ...   ...        ...   
90       brewpub         Sacramento  United States  1158  38.565796   
91         micro           Valencia  United States  1177  34.455562   
92       brewpub    Prescott Valley  United States   155        NaN   
93       brewpub            Oakland  United States  1180  37.806958   
94       brewpub      San Francisco  United States  1182  37.761163   
95         micro    Highlands Ranch  United States  1195  39.565958   
96       brewpub             Golden  United States  1222  39.756211   
97       brewpub      Woodland Park  United States  1227  38.994141   
98         micro           Brighton  United States  1229  39.986429   
99      planning           Trinidad  United States  1333  37.169397   
100     planning               None  United States   207        NaN   
101     contract               None  United States  2217        NaN   
102      brewpub        Little Rock  United States   254  34.742111   
103        micro   Colorado Springs  United States  1250  38.832607   
104        micro             Denver  United States  1279  39.782517   
105        micro             Denver  United States  1296  39.698794   
106        micro         Louisville  United States  1319  39.964883   
107      brewpub       Central City  United States  1337  39.800143   
108        micro         Estes Park  United States  1354  40.371488   
109        micro          San Diego  United States   304  32.701876   
110      brewpub   Glenwood Springs  United States  1370  39.547656   
111      brewpub          Silverton  United States  1373  37.811944   
112      brewpub           Gunnison  United States  1396  38.544623   
113        micro       Fort Collins  United States  1419  40.562046   
114      brewpub          Ev

In [183]:
# Create a subset of the brewery_data_df called brewery_locacation_df that drops those rows that have NaN entries for 'latitude' and 'longitude' - this should reduce the inital list to 4,737
# Follow-up work will go back and get the LAT and LON for those missing locations after getting this inital list of 4,737 to plot in gmaps
brewery_location_df = brewery_data_df.dropna(subset=['latitude', 'longitude'])
brewery_location_df


brewery_type               city        country    id   latitude  \
0           micro         Birmingham  United States     2  33.524521   
1           micro         Tuscaloosa  United States     4  33.198491   
2           micro         Birmingham  United States    44  33.512849   
3           micro         Huntsville  United States    46  34.727752   
4           micro            Wasilla  United States    55  61.575269   
5           micro          Anchorage  United States    76  61.138489   
6           micro             Tucson  United States    94  32.246737   
7        contract         Scottsdale  United States    98  33.497262   
8         brewpub           Chandler  United States   104  33.305346   
9           micro             Tucson  United States   107  32.201608   
10          micro             Tucson  United States   127  32.250495   
11          micro           Williams  United States   141  35.250028   
12          micro   Lake Havasu City  United States   163  34.468974   
13          micro             Peoria  United States   182  33.668774   
14       regional           Chandler  United States   187  33.303244   
15        brewpub           Surprise  United States   198  33.638221   
16          micro            Phoenix  United States   219  33.516633   
17        brewpub          Paragould  United States   225  36.031636   
18          micro  North Little Rock  United States   235  34.759428   
19          micro        Little Rock  United States   247  34.742845   
20          micro      Mountain Home  United States   253  36.332643   
21       planning             Lowell  United States   264  36.255354   
22        brewpub    San Luis Obispo  United States   281  35.246728   
23          micro          San Diego  United States   286  33.023910   
24          micro          Livermore  United States   305  37.678391   
25        brewpub             Auburn  United States   329  38.896121   
26          micro            Anaheim  United States   333  33.799837   
27        brewpub       Rohnert Park  United States   361  38.364970   
28          micro            Lemoore  United States   376  36.301468   
29          micro      San Francisco  United States   385  37.780655   
...           ...                ...            ...   ...        ...   
7986        micro             Dupont  United States  7470  47.107496   
7987        micro             Renton  United States  7472  47.480142   
7988        micro               Kent  United States  7473  47.384495   
7989        micro             Tacoma  United States  7495  47.262094   
7990        micro            Prosser  United States  7511  46.204474   
7991      brewpub          Waitsburg  United States  7527  46.270046   
7992      brewpub            Seattle  United States  7540  47.530340   
8006      brewpub            Bothell  United States  7550  47.763878   
8007     contract            Pacific  United States  7585  47.250134   
8008      brewpub           Winthrop  United States  7596  48.478146   
8009        micro             Tacoma  United States  7604  47.250289   
8010      brewpub            Seattle  United States  7620  47.592009   
8011      brewpub           Bellevue  United States  7637  47.552321   
8012        micro             Yakima  United States  7664  46.605064   
8013      brewpub         Moses Lake  United States  7679  47.128084   
8014   proprietor           Westover  United States  7771  39.636248   
8015        micro          Milwaukee  United States  7797  43.042969   
8016      brewpub         Marshfield  United States  7801  44.659381   
8017     regional          Middleton  United States  7817  43.094896   
8018        micro            Madison  United States  7849  43.050102   
8019      brewpub          Fitchburg  United States  7858  43.018163   
8020     regional          Milwaukee  United States  7879  43.054787   
8021     regional             Monroe  United States  7895  42.598631   
8022      brewpub            Cornell  United 

In [184]:
# Use the existing brewery latitude and longitude to create a heatmap using gmaps
# note gmaps has three possible views, the default is a "TERRAN" but "SATELLITE" or  "HYBRID"
locations = brewery_location_df[['latitude', 'longitude']]

fig = gmaps.figure(map_type='HYBRID')

heatmap_layer = gmaps.heatmap_layer(locations)
fig.add_layer(heatmap_layer)
fig

Figure(layout=FigureLayout(height='420px'))

In [185]:
# list the brewery counts by city using count of postal_codes
breweries_by_state_city_df = brewery_data_df.groupby(['state', 'city'])[['postal_code']].count()
# Sort by state and then city in decending order
breweries_by_state_city_df.sort_values(by='postal_code',ascending=False)

postal_code
state          city                         
Colorado       Denver                     93
California     San Diego                  91
Oregon         Portland                   86
Washington     Seattle                    73
Illinois       Chicago                    65
Texas          Austin                     54
California     San Francisco              39
Minnesota      Minneapolis                38
Ohio           Cincinnati                 35
Indiana        Indianapolis               34
Texas          Houston                    33
New Mexico     Albuquerque                32
North Carolina Charlotte                  31
Ohio           Columbus                   30
Wisconsin      Milwaukee                  30
Ohio           Cleveland                  29
New York       Brooklyn                   29
Colorado       Colorado Springs           28
               Boulder                    27
North Carolina Asheville                  27
Pennsylvania   Pittsburgh                 26
Michigan       Grand Rapids               25
Pennsylvania   Philadelphia               25
Oregon         Bend                       25
Colorado       Fort Collins               24
Missouri       Saint Louis                24
California     Los Angeles                24
Virginia       Richmond                   23
Tennessee      Nashville                  23
Washington     Spokane                    23
...                                      ...
Michigan       Three Rivers                1
               Troy                        1
               Washington                  1
               Waterford                   1
               Watervliet                  1
               Wayland                     1
               West Branch                 1
               Whitehall                   1
               Reed City                   1
               Plainwell                   1
               Lawton                      1
               Paradise                    1
               Lexington                   1
               Lincoln Park                1
               Mackinaw City               1
               Manistee                    1
               Manistique                  1
               Marine City                 1
               Marysville                  1
               Mattawan                    1
               Milan                       1
               Morley                      1
               Munising                    1
               New Hudson                  1
               Newaygo                     1
               Novi                        1
               Oscoda                      1
               Oxford                      1
               PAW PAW                     1
Wyoming        Thermopolis                 1

[3494 rows x 1 columns]

#### Note: There are 3,494 unique cities in the USA with breweries.  
* The top 30 cities (having 23 or more breweries) are listed in the top half (above the ..) and range from Denver, Colorado with 93 breweries to Richmond, Nashville and Spokane all with 23. 
* The bottom 30 cities each have only one brewery - note there are many more cities that have only one brewery
Question: Using Demographic data from US Census, are there any significant differences between cities in the top 30 vs. cities in the bottom 30?

In [124]:
#--------------------------------------------------------------------
# WORK IN PROGRESS
#--------------------------------------------------------------------

# Add GeoJSON to a map to draw a thematic map which displays shading in proportion to the measurement
# being displayed on the map, such as population density or per-capita income.

countries_geojson = gmaps.geojson_geometries.load_geometry('us-states')
fig = gmaps.figure()
geojson_layer = gmaps.geojson_layer(countries_geojson)
fig.add_layer(geojson_layer)
fig

# Scale the number of breweries per state to lie between 0 and 1
min_brewery = min(state2brewery_count.values())
max_brewery = max(state2brewery_count.values())
brewery_range = max_brewery - min_brewery
def calculate_color(brew_index):
    """ 
    Convert the brew_index coefficient to a color
    """
# make brew_index a number between 0 and 1
normalized_brew_index = (brewery_count - min_brewery) / brewery_range
# invert brewery_index so that high inequality gives dark color
inverse_brewery = 1.0 - normalized_brewery_index
# transform the brewery_index coefficient to a matplotlib color
mpl_color = viridis(inverse_gini)
# transform from a matplotlib color to a valid CSS color
gmaps_color = to_hex(mpl_color, keep_alpha=False)
return gmaps_color

Figure(layout=FigureLayout(height='420px'))

## References:

### Literature Review of Craft Brewery Statistics
* https://www.brewersassociation.org/insights/shifting-demographics-among-craft-drinkers/
* https://www.brewersassociation.org/category/insights/
* https://www.brewersassociation.org/statistics/by-state/
* https://www.brewersassociation.org/statistics/national-beer-sales-production-data/
* https://www.brewersassociation.org/statistics/economic-impact-data/
* https://s3-us-west-2.amazonaws.com/brewersassoc/wp-content/uploads/2018/08/Brewers-Association-Economic-Impact-Study-Methodology.pdf
* https://www.brewersassociation.org/statistics/number-of-breweries/
* https://s3-us-west-2.amazonaws.com/brewersassoc/wp-content/uploads/2018/08/State-by-State-Breakdown-2017.pdf
* https://www.brewersassociation.org/statistics/market-segments/
* https://www.brewersassociation.org/government-affairs/laws/barrel-cap-laws/


### Brewery API sites
* https://www.openbrewerydb.org/ # includes a wrapper for making API calls by state, type brewery etc.

### Other API sites considered / used
* https://www.zipcodeapi.com/ #site that translates address, city, state into lat lon coordinates
* https://opencagedata.com #site that translates address, city, state into lat lon coordinates

### Mapping using Google Maps
* https://buildmedia.readthedocs.org/media/pdf/jupyter-gmaps/latest/jupyter-gmaps.pdf

